external/vcm/vcm/data_transform.py

```diff
@@ -0,0 +1,103 @@
+import dataclasses
+from toolz.functoolz import curry
+from typing import Any, Callable, Literal, Mapping, MutableMapping, Sequence, Set
+import xarray as xr
+import vcm
+
+
+DATA_TRANSFORM_REGISTRY: MutableMapping[str, Mapping[str, Any]] = {}
+
+TransformName = Literal[
+    "q1_from_qm_q2", "qm_from_q1_q2", "q1_from_dQ1_pQ1", "q2_from_dQ2_pQ2",
+]
+
+
+@curry
+def register(
+    name: str,
+    inputs: Sequence[str],
+    outputs: Sequence[str],
+    func: Callable[..., xr.Dataset],
+):
+    if name in DATA_TRANSFORM_REGISTRY:
+        raise ValueError(f"Function {name} has already been added to registry.")
+    DATA_TRANSFORM_REGISTRY[name] = {
+        "func": func,
+        "inputs": inputs,
+        "outputs": outputs,
+    }
+    return func
+
+
+@register("qm_from_q1_q2", ["Q1", "Q2"], ["Qm"])
+def qm_from_q1_q2(ds):
+    ds["Qm"] = vcm.moist_static_energy_tendency(ds["Q1"], ds["Q2"])
+    return ds
+
+
+@register("q1_from_qm_q2", ["Qm", "Q2"], ["Q1"])
+def q1_from_qm_q2(ds):
+    ds["Q1"] = vcm.temperature_tendency(ds["Qm"], ds["Q2"])
+    return ds
+
+
+@register("q1_from_dQ1_pQ1", ["dQ1", "pQ1"], ["Q1"])
+def q1_from_dQ1_pQ1(ds):
+    ds["Q1"] = ds["dQ1"] + ds["pQ1"]
+    return ds
+
+
+@register("q2_from_dQ2_pQ2", ["dQ2", "pQ2"], ["Q2"])
+def q2_from_dQ2_pQ2(ds):
+    ds["Q2"] = ds["dQ2"] + ds["pQ2"]
+    return ds
+
+
+@dataclasses.dataclass
+class DataTransformConfig:
+    name: TransformName
+    kwargs: dict = dataclasses.field(default_factory=dict)
+
+
+@dataclasses.dataclass
+class DataTransform:
+    config: DataTransformConfig
+
+    def apply(self, ds: xr.Dataset) -> xr.Dataset:
+        func = DATA_TRANSFORM_REGISTRY[self.config.name]["func"]
+        ds = func(ds, **self.config.kwargs)
+        return ds
+
+    @property
+    def input_variables(self) -> Sequence[str]:
+        return DATA_TRANSFORM_REGISTRY[self.config.name]["inputs"]
+
+    @property
+    def output_variables(self) -> Sequence[str]:
+        return DATA_TRANSFORM_REGISTRY[self.config.name]["outputs"]
+
+
+@dataclasses.dataclass
+class ChainedDataTransform:
+    config: Sequence[DataTransformConfig]
```

Comment:
suggestion: `transforms` for the name, or something else that describes the intent of the config? we already have "config" in our argo yamls to indicate the training config, as it stands this PR would put the "config" key at three levels of that yaml.

Reply:
That sounds good. Although I actually avoided using the `ChainedDataTransform` in the top-level configs. I used `Sequence[DataTransform]` directly there, otherwise we end up with an unnecessary single-itemed dict which just confuses things IMO. It does mean I have to do some custom handling for dumping/loading (e.g. see the `TransformedPredictor` dump/load methods).

Reply:
Let me know if you have other ideas to handle this.